types: PrivValidator.Save should first invoke os.MkdirAll

odeke-em opened the pull request:
Fixes https://github.com/tendermint/tendermint/issues/1290

We invoked cmn.WriteFileAtomic without first checking if
the parent directory existed yet cmn.WriteFileAtomic doesn't
do it first.

This change ensures that we invoke os.MkdirAll first.

With it, we now get
```shell
tendermint testnet
Successfully initialized 4 node directories
```
instead of panicking.

<!-- Thanks for filing a PR! Before hitting the button, please check the following items.-->

* [X] Updated all relevant documentation in docs
* [X] Updated all code comments where relevant
* [X] Wrote tests
* [ ] Updated CHANGELOG.md

melekes commented on types/priv_validator.go:
```diff
@@ -225,7 +227,13 @@ func (privVal *PrivValidatorFS) save() {
-	err = cmn.WriteFileAtomic(privVal.filePath, jsonBytes, 0600)
+	outFile := privVal.filePath
+	dir := filepath.Dir(outFile)
+	if err := os.MkdirAll(dir, 0755); err != nil {
```
I think permission should also be 0600

odeke-em replied:
0600 will only allow for Read and Writes for the owner of that directory, if we opened up another terminal and tried to get content out of that directory, we'd get "Permission denied" as 6 for the owner would mean they can't invoke execute which seems to be needed for MkdirAll, let's meet in the middle, how about `0700`? Still then I even wonder if other users in the same group will be able to use the file as they'll have `0` for their access

xla commented on types/priv_validator.go:
```diff
@@ -225,7 +227,14 @@ func (privVal *PrivValidatorFS) save() {
+	// See Issue https://github.com/tendermint/tendermint/issues/1290
+	dir := filepath.Dir(outFile)
+	if err := os.MkdirAll(dir, 0700); err != nil {
+		cmn.PanicCrisis(err)
```
AFAIK `Panic` methods are deprecated, we use blank `panic`s now.

odeke-em replied:
I see, thank you. I was using the surrounding code as the guide. However, I'll change that up.

xla commented on types/priv_validator_test.go:
```diff
@@ -253,3 +255,15 @@ func clipToMS(t time.Time) time.Time {
+
+	pPath := filepath.Join(saveDir, "out-file", "tf.json")
+	pv := &PrivValidatorFS{filePath: pPath}
+	pv.Save()
```
While it is a regression test, the behaviour we want to ensure is that the file is saved even in nested, non-existing directory structure. Would rename the test accordingly and test that the file with the right content is created for extra verification.

odeke-em replied:
Gotcha, thanks for that @xla, in deed that adds extra verification. In the past I've had pushback whenever I've pushed up rigorous tests, but am glad to get support for such, from you so I'll implement it :)

odeke-em replied:
Great stuff @xla recommending the rigorous validation, me likey!

melekes commented on types/priv_validator.go:
```diff
@@ -225,7 +227,14 @@ func (privVal *PrivValidatorFS) save() {
+	// We need to ensure that the parent directory exists.
+	// See Issue https://github.com/tendermint/tendermint/issues/1290
+	dir := filepath.Dir(outFile)
+	if err := os.MkdirAll(dir, 0700); err != nil {
```
Should we move this into `cmn.WriteFileAtomic`? The flow is

1 <...> - 2 <priv_validator.go> - 3 <common.go>

So we should either create the directory at (1) or (3). (2) feels like a halfway.

xla replied:
:+1:

odeke-em replied:
Great point @melekes, in that case I'd rather do it in 3) cmn.WriteFileAtomic. I didn't do so initially because I assumed changing the implementation of cmn.WriteFileAtomic was off-limits.

odeke-em replied:
Okay now I recall why I put this here, doing MkdirAll in cmn.WriteFileAtomic is quite complex:
Atomic meaning "all or nothing" would require that I have to check the:
a) The least none existent path e.g if we have a/b/c/d/e/f/g/txt.json, if directory b doesn't exist, I'll have to find out firstly what doesn't exist to clean it up if the operation fails
b) Remove the least existent path

I am not sure I'd want to be introducing such  a change in common at a time like this close to launch and moreover when my dev bandwidth is running out. For those reasons, I'll leave this here, someone else can
later move that check into WriteFileAtomic since that had never come up anyways.

melekes replied:
ok, I tend to agree

odeke-em commented:
Thank you @melekes and @xla for the reviews, I've addressed all the feedback except for moving the os.MkdirAll out of where it is currently, because, I believe that doing so will expend currently unavailable dev cycles.

xla commented:
Rebased on latest `develop`.

ebuchman commented:
Thanks folks but the solution is much simpler: https://github.com/tendermint/tendermint/pull/1318

PrivVal can just require that the folder is already set up - it's the callers duty to initialize. 

This failure is a vestige of the introduction of the config/data folders and the lack of testing the `testnet` command ...
